### PR TITLE
✨New Feature: Remove Gene Tracks from Oncoprint

### PR DESCRIPTION
**Pull request by SAHU-01**
Fix cBioPortal/cbioportal#10612 (see https://help.github.com/en/articles/closing-issues-using-keywords)

Describe changes proposed in this pull request:
- Make gene tracks removable by setting `removable:true` in DeltaUtils.ts
- Remove gene tracks from track menu in oncoprint UI, by updating URL `gene_list` 

## Checks
- [ ] Has tests or has a separate issue that describes the types of test that should be created. If no test is included it should explicitly be mentioned in the PR why there is no test.
- [ ] The commit log is comprehensible. It follows [7 rules of great commit messages](http://chris.beams.io/posts/git-commit/). For most PRs a single commit should suffice, in some cases multiple topical commits can be useful. During review it is ok to see tiny commits (e.g. Fix reviewer comments), but right before the code gets merged to master or rc branch, any such commits should be squashed since they are useless to the other developers. Definitely avoid [merge commits, use rebase instead.](http://nathanleclaire.com/blog/2014/09/14/dont-be-scared-of-git-rebase/)
- [ ] Is this PR adding logic based on one or more **clinical** attributes? If yes, please make sure validation for this attribute is also present in the data validation / data loading layers (in backend repo) and documented in [File-Formats Clinical data section](https://github.com/cBioPortal/cbioportal/blob/master/docs/File-Formats.md#clinical-data)!

## Any screenshots or GIFs?
If this is a new visual feature please add a before/after screenshot or gif
here with e.g. [Giphy CAPTURE](https://giphy.com/apps/giphycapture) or [Peek](https://github.com/phw/peek)
Before:
image

After:
image

## Notify reviewers
@inodb @alisman 
Read our [Pull request merging
policy](../CONTRIBUTING.md#pull-request-merging-policy). It can help to figure out who worked on the
file before you. Please use `git blame <filename>` to determine that
and notify them either through slack or by assigning them as a reviewer on the PR

### CODE REVIEW

File: src/shared/components/oncoprint/ResultsViewOncoprint.tsx [hunk 1]
```diff
@@ -364,6 +366,23 @@ export default class ResultsViewOncoprint extends React.Component<
         ) as ClinicalTrackConfigMap;
     }
 
+    @computed get selectedGeneticTrackConfig(): GeneticTrackConfigMap {
+        let geneticTracks: GeneticTrackConfig[] | undefined = this.props.store
+            .pageUserSession.userSettings?.geneticlist;
+        if (geneticTracks) {
+            const userSettingsTracksMap = {} as GeneticTrackConfigMap;
```
Comment (alisman) [inline src/shared/components/oncoprint/ResultsViewOncoprint.tsx:373]: minor quibble but try to use funtional approach:

const userSettingsTrackMap = genetTracks.map(...)

File: src/shared/components/oncoprint/ResultsViewOncoprint.tsx [hunk 2]
```diff
@@ -364,6 +366,23 @@ export default class ResultsViewOncoprint extends React.Component<
...
+            geneticTracks.forEach(g => (userSettingsTracksMap[g.stableId] = g));
+            return userSettingsTracksMap;
+        }
+
+        geneticTracks = [];
+
+        _.forEach(this.props.store.genes.result, attr => {
```
Comment (alisman) [inline src/shared/components/oncoprint/ResultsViewOncoprint.tsx:380]: same here.   will be more readable

**Review by inodb — COMMENTED** [review]
@SAHU-01 Thanks again for working on this!

- [x] There's a small bug when removing the last gene. Gives an error page. Not sure what the behavior should be. Maybe the remove track button should be disabled when there's only one gene?

**Comment by SAHU-01** [comment]
@inodb thank you for pointing out the bug, I had missed this edge case, as per your suggestion now, if there is more than 1 gene track present we can delete the tracks, else the "Remove Track" option is disabled when only 1 gene track is present.

**Review by inodb — APPROVED** [review]
@SAHU-01 Thanks so much for fixing this so quickly!

**Review by inodb — CHANGES_REQUESTED** [review]
@SAHU-01 sorry @alisman found some more bugs:

- [x] When using the merged track feature, see e.g. [this query](https://deploy-preview-4861--cbioportalfrontend.netlify.app/results/oncoprint?cancer_study_list=coadread_tcga_pub&cancer_study_id=coadread_tcga_pub&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=coadread_tcga_pub_mutations&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=coadread_tcga_pub_gistic&Z_SCORE_THRESHOLD=2.0&case_set_id=coadread_tcga_pub_nonhypermut&gene_list=%255B%2520%2522RAS%252FRAF%2522%2520KRAS%2520NRAS%2520BRAF%2520%255D%2520PTEN&gene_set_choice=user-defined-list&RPPA_SCORE_THRESHOLD=2.0&profileFilter=mutations%2Cgistic&geneset_list=%20&tab_index=tab_visualize&Action=Submit). All the genes become unmerged  when one gene is removed. The expected behavior would be that 1) when removing the merged track, the entire merged track is removed and 2) when removing the one gene, the existing merged track stays
- [ ] Other OQL modifiers also get removed when removing a track, see e.g. [this query](https://deploy-preview-4861--cbioportalfrontend.netlify.app/results/oncoprint?cancer_study_list=bladder_columbia_msk_2018&Z_SCORE_THRESHOLD=2.0&RPPA_SCORE_THRESHOLD=2.0&profileFilter=mutations%2Ccna&case_set_id=bladder_columbia_msk_2018_cnaseq&gene_list=BRAF%253A%2520MUT%2520%253D%2520V600E%250ANRAS%250AKRAS&geneset_list=%20&tab_index=tab_visualize&Action=Submit)
- [ ] Same issue for `DATATYPES`, see [this query](https://deploy-preview-4861--cbioportalfrontend.netlify.app/results/oncoprint?cancer_study_list=acc_tcga_pan_can_atlas_2018&Z_SCORE_THRESHOLD=2.0&RPPA_SCORE_THRESHOLD=2.0&profileFilter=mutations%2Cstructural_variants%2Cgistic%2Crna_seq_v2_mrna_median_Zscores&case_set_id=acc_tcga_pan_can_atlas_2018_3way_complete&gene_list=DATATYPES%253A%2520AMP%2520GAIN%2520HOMDEL%2520EXP%2520%253E%25201.5%2520EXP%2520%253C%2520-1.5%253B%2520CDKN2A%2520MDM2%2520TP53&geneset_list=%20&tab_index=tab_visualize&Action=Submit)

The documentation for OQL can be found at https://cbioportal.org/oql if helpful

Thanks in advance for looking at it!

**Comment by SAHU-01** [comment]
@inodb and @alisman regarding the bug in **_merged track feature_**, I've fixed it in the above commit and regarding **_OQL modifiers and DATATYPES_** issue, though I've found the cause yet I need some time to explore the best possible fix for it, regardless I'll fix it by tomorrow!

**Comment by SAHU-01** [comment]
@inodb and @alisman I've fixed all the above-listed bugs, do let me know if there are any more bugs or if the code/logic needs more optimization!

**Comment by alisman** [comment]
@SAHU-01 for OQL modifiers, I still see the problem IF i remove the first track (BRAF) though it works for subsequent tracks

**Comment by SAHU-01** [comment]
> @SAHU-01 for OQL modifiers, I still see the problem IF i remove the first track (BRAF) though it works for subsequent tracks

@alisman on removing (BRAF), though the modifier (**MUT AMP HOMDEL**) of the other 2 tracks are not visible, but they still work perfectly, the thing is,  when any track is removed it's modifiers are also removed from URL but in this case, in the URL the default modifier (MUT AMP HOMDEL) are not present so after (BRAF) is removed, even though the default modifier(MUT AMP HOMDEL) works as usual they are not visible in UI after deletion.

You can see it in this console where on deleting each track the sublabel shared details about the OQL modifier present
image

When you try removing one of the other 2 tracks first, it's works perfectly because then each track and that track's sublabel is independently pushed to URL like below, I thought specifically pushing sublabels with remaining tracks, when only default tracks remain won't be wise, but if you suggest to push sublabels in case of defaults then I can modify accordingly.
image

**Comment by alisman** [comment]
@SAHU-01 thanks for your careful analysis here.  I am going to need to think about the right thing to do and discuss with product team.  These are probably edge cases we're worried about so maybe it's ok not worry about them.  Stay tuned.

**Pull request merged**